src/string.cr

```diff
@@ -3672,74 +3712,83 @@ class String
   # "3.14IsPi".underscore                                       # => "3.14_is_pi"
   # "InterestingImage".underscore(Unicode::CaseOptions::Turkic) # => "ınteresting_ımage"
   # ```
-  def underscore(options : Unicode::CaseOptions = Unicode::CaseOptions::None)
+  def underscore(options : Unicode::CaseOptions = :none) : String
+    String.build(bytesize + 10) { |io| underscore io, options }
+  end
+
+  # Writes an underscored version of `self` to the given *io*.
+  #
+  # ```
+  # io = IO::Memory.new
+  # "DoesWhatItSaysOnTheTin".underscore io
+  # io.to_s # => "does_what_it_says_on_the_tin"
+  # ```
+  def underscore(io : IO, options : Unicode::CaseOptions = :none) : Nil
     first = true
     last_is_downcase = false
     last_is_upcase = false
     last_is_digit = false
-    mem = nil
+    mem : Char? = nil
```

Comment:
Why the type restriction?

Reply:
Make it more clear to the next person what it is.  I suppose I could have renamed it to `char` or something, but this was the least invasive solution.